File: Day221/StackUsingQueues.java

class MyStack {
    Queue<Integer> queue1;
    Queue<Integer> queue2;
    int top = -1;
    public MyStack() {
        queue1 = new LinkedList<>();
        queue2 = new LinkedList<>();
    }
    
    public void push(int x) {
        queue1.offer(x);
        top = x;
    }
    
    public int pop() {
        int to = -1;
        if (!empty()) {
            while(queue1.size()!=1) {
                queue2.offer(queue1.poll());
            }
             to = queue1.poll();
            while(!queue2.isEmpty()) {
               top = queue2.poll();
                queue1.offer(top);
            }
        }   
        return to;
    }
    
    public int top() {
        return top;
    }
    
    public boolean empty() {
        return queue1.size()==0;
    }
}

File: Day221/StackUsingQueues.go
package main

import "container/list"

type MyStack struct {
	queue1 *list.List
	queue2 *list.List
	top    int
}

func Constructor() MyStack {
	return MyStack{
		queue1: list.New(),
		queue2: list.New(),
		top:    -1,
	}
}

func (stack *MyStack) Push(x int) {
	stack.queue1.PushBack(x)
	stack.top = x
}

func (stack *MyStack) Pop() int {
	to := -1
	if !stack.Empty() {
		for stack.queue1.Len() != 1 {
			elem := stack.queue1.Front()
			stack.queue1.Remove(elem)
			stack.queue2.PushBack(elem.Value)
		}
		toElem := stack.queue1.Front()
		stack.queue1.Remove(toElem)
		to = toElem.Value.(int)
		for stack.queue2.Len() != 0 {
			topElem := stack.queue2.Front()
			stack.queue2.Remove(topElem)
			stack.top = topElem.Value.(int)
			stack.queue1.PushBack(stack.top)
		}
	}
	return to
}

func (stack *MyStack) Top() int {
	return stack.top
}

func (stack *MyStack) Empty() bool {
	return stack.queue1.Len() == 0
}

func main() {
	stack := Constructor()
	stack.Push(1)
	stack.Push(2)
	fmt.Println(stack.Top()) // Output: 2
	fmt.Println(stack.Pop()) // Output: 2
	fmt.Println(stack.Empty()) // Output: false
}
